File: app/src/main/java/com/mmutert/freshfreezer/data/ItemDao.java
package com.mmutert.freshfreezer.data;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;
import androidx.room.Update;

import java.util.List;

@Dao
public abstract class ItemDao {

    public static final int TRUE = 1;
    public static final int FALSE = 0;

    // ============================== Items ====================================

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract long insertItem(FrozenItem item);

    @Query("SELECT * FROM items")
    public abstract LiveData<List<FrozenItem>> getAllItems();

    @Delete
    public abstract void deleteItem(FrozenItem item);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    public abstract void updateFrozenItem(FrozenItem item);


    // ============================ Notifications =================================

    @Query("Select * from notifications")
    public abstract LiveData<List<ItemNotification>> getAllNotificationsLiveData();

    @Query("Select * from notifications where item_id = :itemId")
    public abstract LiveData<List<ItemNotification>> getAllNotificationsLiveData(long itemId);

    @Query("Select * from notifications where item_id = :itemId")
    public abstract List<ItemNotification> getAllNotifications(long itemId);

    public LiveData<List<ItemNotification>> getAllNotificationsLiveData(FrozenItem item) {
        return getAllNotificationsLiveData(item.getId());
    }

    public List<ItemNotification> getAllNotifications(FrozenItem item) {
        return getAllNotifications(item.getId());
    }

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    public abstract void addNotification(ItemNotification notification);

    @Delete
    public abstract void deleteNotification(ItemNotification notification);


    // ========================= Combined Items and Notifications ============================

    @Transaction
    @Query("Select * from items where id = :id")
    public abstract LiveData<ItemAndNotifications> getItemAndNotificationsLiveData(long id);

    @Transaction
    @Query("Select * from items where id = :id")
    public abstract ItemAndNotifications getItemAndNotifications(long id);

    @Transaction
    @Query("Select * from items")
    public abstract LiveData<List<ItemAndNotifications>> getItemAndNotificationsList();
}

File: app/src/main/java/com/mmutert/freshfreezer/data/ItemDao.kt
package com.mmutert.freshfreezer.data

import androidx.lifecycle.LiveData
import androidx.room.*

@Dao
interface ItemDao {
    // ============================== Items ====================================
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertItem(item: FrozenItem): Long

    @get:Query("SELECT * FROM items")
    val allItems: LiveData<List<FrozenItem>>

    @Delete
    suspend fun deleteItem(item: FrozenItem?)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    suspend fun updateFrozenItem(item: FrozenItem?)

    // ============================ Notifications =================================
    @get:Query("Select * from notifications")
    val allNotificationsLiveData: LiveData<List<ItemNotification>>

    @Query("Select * from notifications where item_id = :itemId")
    fun getAllNotificationsLiveData(itemId: Long): LiveData<List<ItemNotification>>

    @Query("Select * from notifications where item_id = :itemId")
    fun getAllNotifications(itemId: Long): List<ItemNotification>

    fun getAllNotificationsLiveData(item: FrozenItem): LiveData<List<ItemNotification>> {
        return getAllNotificationsLiveData(item.id)
    }

    fun getAllNotifications(item: FrozenItem): List<ItemNotification> {
        return getAllNotifications(item.id)
    }

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun addNotification(notification: ItemNotification?)

    @Delete
    suspend fun deleteNotification(notification: ItemNotification?)

    // ========================= Combined Items and Notifications ============================
    @Transaction
    @Query("Select * from items where id = :id")
    fun getItemAndNotificationsLiveData(id: Long): LiveData<ItemAndNotifications>

    @Transaction
    @Query("Select * from items where id = :id")
    fun getItemAndNotifications(id: Long): ItemAndNotifications

    @get:Transaction
    @get:Query("Select * from items")
    val itemAndNotificationsList: LiveData<List<ItemAndNotifications>>

}